contrib/tinks/src/client/client.nit

```diff
@@ -0,0 +1,433 @@
+			var server = new RemoteServer(server_config)
+
+			# Connect then complete handshake
+			assert server.connect else print_error "Connection to server failed with {server.socket.last_error or else "none"}"
+			assert server.handshake else print_error "Handshake with server failed"
+
+			# Download and setup remote game
+			var context = new RemoteGameContext(server)
+			context.setup
+
+			return context
+		end
+	end
+
+	# `Tank` of the local player, if any
+	fun local_tank: nullable Tank
+	do
+		# FIXME use a ? to one line this
+		var local_player = context.local_player
+		if local_player == null then return null
+		return local_player.tank
+	end
+
+	# Square of the minimum distance from the tank for an object to be "far"
+	#
+	# This value influences which sounds are heard,
+	# the strength of vibrations and
+	# whether an arrow points to a far unit
+	private var far_dist2 = 2000.0
+
+	# Tank tracks tracks on the ground
+	#
+	# TODO use particles or at least optimize drawing
+	var tracks = new List[Couple[Pos, Float]]
+
+	redef fun frame_core(display)
+	do
+		var clock = new Clock
+
+		var turn = context.do_turn
+		sys.perfs["do_turn"].add clock.lapse
+
+		# Draw
+
+		# Update camera
+		if down_keys.has("left") then camera.dx -= 1.0
+		if down_keys.has("right") then camera.dx += 1.0
+		if down_keys.has("up") then camera.dy -= 1.0
+		if down_keys.has("down") then camera.dy += 1.0
+
+		var local_tank = local_tank
+		if local_tank != null then
+			var tank_speed = local_tank.direction_forwards*local_tank.rule.max_speed
+			tank_speed = tank_speed.min(0.5).max(-0.5)
+
+			var prop_pos = local_tank.pos + local_tank.heading.to_vector(tank_speed * 16.0)
+			var old_pos = camera.center(display)
+			var half = old_pos.lerp(prop_pos, 0.02)
+
+			camera.center_on(display, new Pos(half.x, half.y))
+		end
+
+		# Grass
+		display.clear(0.0, 0.45, 0.0)
+
+		# Past tank tracks
+		for track in tracks do
+			var pos = track.first.to_screen(camera)
+			display.blit_rotated(assets.drawing.track, pos.x, pos.y, track.second)
+		end
+
+		# Past blast sites
+		for blast in context.game.world.blast_sites do
+			var pos = blast.to_screen(camera)
+			display.blit_centered(assets.drawing.blast, pos.x, pos.y)
+		end
+
+		# Terrain features
+		var tl = (new ScreenPos(0.0, 0.0)).to_logic(camera)
+		var br = (new ScreenPos(display.width.to_f, display.height.to_f)).to_logic(camera)
+		for x in [tl.x.floor.to_i .. br.x.ceil.to_i] do
+			for y in [tl.y.floor.to_i .. br.y.ceil.to_i] do
+				var feature = context.game.world[x, y]
+				if feature != null then
+					var pos = feature.pos.to_screen(camera)
+					var image = feature.rule.images[feature.image_index]
+					display.blit_rotated(image, pos.x, pos.y, feature.angle)
+				end
+			end
+		end
+
+		# Tanks
+		for tank in context.game.tanks do
+			# Add random tracks
+			if (tank.direction_heading != 0.0 and 40.rand == 0) or
+			   (tank.direction_forwards != 0.0 and 100.rand == 0) then
+
```

Comment:
for the same number of lines, an alternative style could be to use the `then` to separate blocks:

~~~nit
if (tank.direction_heading != 0.0 and 40.rand == 0) or
        (tank.direction_forwards != 0.0 and 100.rand == 0)
then
        tracks.add new Couple[Pos, Float](tank.pos, tank.heading)
        if tracks.length > 1000 then tracks.shift
end
~~~

with an additional line, you can also allign the conditions

~~~nit
if
        (tank.direction_heading != 0.0 and 40.rand == 0) or
        (tank.direction_forwards != 0.0 and 100.rand == 0)
then
        tracks.add new Couple[Pos, Float](tank.pos, tank.heading)
        if tracks.length > 1000 then tracks.shift
end
~~~